toolchain: Fixed ARM toolchain compilation for gcc 8.x

borkra opened the pull request:
Set default GCC ARM CPU and FPU architectures, to resolve kernel compilation failures under gcc 8.x

borkra committed:
toolchain: Fixed ARM toolchain compilation for gcc 8.x

Set default GCC ARM CPU and FPU architectures,
to resolve kernel compilation failures under gcc 8.x

Signed-off-by: Boris Krasnovskiy <borkra@gmail.com>

wongsyrone commented:
Then the toolchain is bounded to the specific CPU and FPU

borkra commented:
No it's not, nothing is bound. 
I just changed just the compiler default settings (settings that active when nothing specified on the command line). All of the defaults could be overwritten by -m option(s) on the command line.
We have to have FPU and CPU defaults that are compatible with hard float default, tp avoid compilation issues.

The problem was that we specify hard float default, but did not select FPU and CPU default that supported hard float, as a result kernel compilation was failing on gcc 8.

In general it does not look like OpenWRT uses gcc feature of being able to have single  toolchain for all ARM CPUs. It prefixes build directory with CPU and FPU type, and that causes tool chain recompile for every CPU/FPU combination.

chunkeey commented:
@borkra 

https://github.com/openwrt/openwrt/pull/1203

Are these the same errors you are talking about?

> --with-fpu=$(word 2, $(subst +, ",$(CONFIG_CPU_TYPE)))

I guess I have to try this first, but there are targets (bcm53xx) that do not have any fpu and hence don't set CPU_SUBTYPE, so CONFIG_CPU_TYPE looks like this: ```CONFIG_CPU_TYPE="cortex-a9"```

Do you know if --with-fpu="" is supported by gcc 8? or does this need a extra if-empty-then-none check?

borkra commented:
@chunkeey 
> #1203
> Are these the same errors you are talking about?
Yes, these are the errors.

> --with-fpu=$(word 2, $(subst +, ",$(CONFIG_CPU_TYPE)))
> 
> I guess I have to try this first, but there are targets (bcm53xx) that do not have any fpu and hence don't set CPU_SUBTYPE, so CONFIG_CPU_TYPE looks like this: CONFIG_CPU_TYPE="cortex-a9"

I guess you are correct --with-fpu need to be set only if --hard-float is selected, I will fix that.

> Do you know if --with-fpu="" is supported by gcc 8? or does this need a extra if-empty-then-none check?
Yes, --with-fpu is supported on gcc 8 - it is the most important part of the fix, When we specify --hard-float, FPU need to be actually be supporting one.

I tried on mvebu, it's cortex-a9  with FPU, with gcc 8.3 and 7.4

diizzyy commented:
Great job!

fuqiang03 commented:
![QQ截图20190317174858](https://user-images.githubusercontent.com/26851715/54488485-2e1aee80-48dd-11e9-8e14-b9cd98dab08f.png)
![QQ截图20190317174930](https://user-images.githubusercontent.com/26851715/54488488-33783900-48dd-11e9-96d4-b8ba9af25cc5.png)

borkra commented:
@fuqiang03 I tried glibc build and I do not see the same errors. 

With this patch, you have to make sure that no artifacts remaining from the previous build.

Please make sure build_dir and staging_dir directories are completely empty before running build with this patch.

diizzyy commented:
Why the change from mcpu to mtune? From what I can tell from GCC documentation mcpu is to be preferred over mtune.

borkra commented:
@diizzyy Please read the commit description.
For ARM that is not the case, for ARM mtine is better then mcpu, again depends on ones objectives.

Here the explanation from commit description:

The root of the issue here is that glibc support "ifunc", while musl does not.
libatomic inside gcc when ifunc support detected adds -march flag,
which conflicts with -mcpu flag set by the target.mk
The issue is -mcpu for ARM is very strict flag, that is
forcing CPU type with no variations permitted of any kind.
As such the solution replace -mcpu, with -mtune, which does the same thing,
but less strict and permits override with the compatible architecture

fuqiang03 commented:
![QQ截图20190318224355](https://user-images.githubusercontent.com/26851715/54538723-b5e12580-49cf-11e9-8f68-28146371628b.png)
Use the command "make dirclean"
But the problem remains.

borkra commented:
@fuqiang03

Make dirclean sometimes leave stuff behind.

I do not see the issue on my local system, everything compiles fine.

Thus the questions:

Are you sure you are building plain openwrt master build with no modifications?

Please provide .config from your build so I could duplicate your configuration.

borkra commented:
@fuqiang03

Make dirclean sometimes leave stuff behind.

I do not see the issue on my local system, everything compiles fine.

Thus the questions:

Are you sure you are building plain openwrt master build with no modifications?

Please provide .config from your build so I could duplicate your configuration.

borkra commented:
@fuqiang03 

I noticed you trying to compile OpenWRT in Thumb mode.

By default OpenWRT does not support this, you must have added this on top of standard OpenWRT yourself.

As such this is way outside of scope for this pull request.
Although I agree it would be cool feature to add, at some later time.

chunkeey commented:
@borkra 

I found @fuqiang03 screens on https://github.com/openwrt/packages/issues/8419 from three days ago. I'm not sure, but It looks like this is an existing problem and that he wants to piggy-pack this onto you.

borkra commented:
@chunkeey 

I would greatly appreciate if you would take the 2nd commit it's a real problem that needs to be fixed.

diizzyy commented:
@borkra 
That sounds more like a broken behaviour by glibc? I'll try to have a look at what archarmlinux does

borkra commented:
@diizzyy 

Glibc has very little to do with this, ifunc is a valid feature, and all option selection  happens in gcc, not glibc
This issue is old and well known too, you can see here, or I can add 20 more links:

https://stackoverflow.com/questions/41934764/gnu-arm-warning-mcpu-cortex-r5-conflicts-with-march-armv7-r-switch?rq=1

borkra commented:
@diizzyy

the other option to solve this is to disable multi-arch in glibc. As we are building monolithic images I am not sure why multi-arch would be needed.

The correct thing in my view is to do both changes, our compiler settings, should not preclude multi-arch, but we do not need it, so disable.

fuqiang03 commented:
@borkra 

[.config.zip](https://github.com/openwrt/openwrt/files/2983021/default.config.zip)
Source code not modified

fuqiang03 commented:
![QQ截图20190319211000](https://user-images.githubusercontent.com/26851715/54608330-6fe99780-4a8b-11e9-93b9-ea10338c0060.png)
Compiled by GCC 7.3
No other modifications

chunkeey commented:
> @chunkeey
>
>I would greatly appreciate if you would take the 2nd commit it's a real problem that needs to be fixed.

I would really like both ;) But I was surprised by the glibc and now I have to setup a test for it. At least with the config it should be possible to see what's going wrong. Please stay tuned.

borkra commented:
@fuqiang03

I do not observe any of the issues you are reporting everything works just fine for me, even with your config.

Statement from you screenshot" "selected processor does not support 'xxxx' in Thumb mode"

Makes me believe that you running modified build. OpenWRT does not support or enable Thumb mode.

fuqiang03 commented:
@borkra 
The same problem can occur with unmodified source code

borkra commented:
@fuqiang03

Not that I can see, everything works fine

chunkeey commented:
@borkra 
I think @fuqiang03 is here because he wants to report a bug (see his other thread on packages in https://github.com/openwrt/packages/issues/8419 ), but not necessarily for your patches.

fuqiang03 commented:
![QQ截图20190320220543](https://user-images.githubusercontent.com/26851715/54690483-772ca600-4b5c-11e9-858f-935d82ff562c.png)
All the methods have been used and still make mistakes.
Gcc8.3 still needs time

chunkeey commented:
> All the methods have been used and still make mistakes.

Really? But were is the "-mcpu" coming from? In @borkra  second patch the openwrt's mcpu is replaced by "-mtune". https://github.com/openwrt/openwrt/pull/1913/commits/e8c170b64de13a33266a1ff9a7a0ce1017003c51

In the mean time, I "dug up" a similar issue from crosstool-ng project:
https://github.com/crosstool-ng/crosstool-ng/issues/963

which brings up a GCC thread that states:
https://gcc.gnu.org/ml/gcc-help/2018-05/msg00120.html

and the response in:
https://gcc.gnu.org/ml/gcc-help/2018-05/msg00124.html

|I think your problem is that you're trying to override
|C(XX)FLAGS_FOR_TARGET with options that contain a CPU.  That means that
|the settings from that variable are conflicting with those added by the
|makefile - so it's nothing to do with the configure's --with-cpu option.

crosstools-ng fixed it by getting rid of the cpu,tune,arch in the CFLAGS (see subject).

https://github.com/crosstool-ng/crosstool-ng/pull/978/commits/07ec87f14f2bb078235a7db25d4a923e9319f53c

So what if we do the same? And just filter -mcpu and -mfpu (?maybe -mfloat,-mtune and -march as well?) flags that get passed to the glibc build options.

```
diff --git a/toolchain/glibc/common.mk b/toolchain/glibc/common.mk
index 93aab8efd2..41e2357214 100644
--- a/toolchain/glibc/common.mk
+++ b/toolchain/glibc/common.mk
@@ -43,11 +43,13 @@ endif
 # -Os miscompiles w. 2.24 gcc5/gcc6
 # only -O2 tested by upstream changeset
 # "Optimize i386 syscall inlining for GCC 5"
+# -mcpu/-mfpu is problematic since they can clash with -march, that's
+# why we let the toolchain decide what it wants.
 GLIBC_CONFIGURE:= \
 	unset LD_LIBRARY_PATH; \
 	BUILD_CC="$(HOSTCC)" \
 	$(TARGET_CONFIGURE_OPTS) \
-	CFLAGS="-O2 $(filter-out -Os,$(call qstrip,$(TARGET_CFLAGS)))" \
+	CFLAGS="-O2 $(filter-out -Os -mcpu=% -mfpu=%,$(call qstrip,$(TARGET_CFLAGS)))" \
 	libc_cv_slibdir="/lib" \
 	use_ldconfig=no \
 	$(HOST_BUILD_DIR)/$(GLIBC_PATH)configure \
```

borkra commented:
@chunkeey

If we want to keep ability to configure compiler through -m options we replacing -mcpu with -mtune will do the trick. 

The issue -mthumb  could be fixed. I suspect the cause of it is:  we do not push TARGET_ASFLAGS to the glibc, in addition TARGET_ASFLAGS and TARGET_CXXFLAGS, missing some important architectural options.

chunkeey commented:
@borkra (I think one of your comments got deleted?!)

As for whenever to use -mtune or -mcpu. There's an recent article from ARM that goes a bit into the details about these compiler flags.

https://community.arm.com/developer/tools-software/tools/b/tools-software-ides-blog/posts/compiler-flags-across-architectures-march-mtune-and-mcpu

I know people won't read it, but it comes with the following recommendation:
"TL;DR: Whenever possible use only -mcpu=native. Avoid -march and -mtune."

So I would suggest to "do what we have been told" and stick with --with-cpu & --with-fpu (patch 1), but filter the mcpu out of the special toolchain cases since I have to trust that GCC and glibc to know what they are doing. Because otherwise this could be stuck until the problem gets propagated to the GCC and glibc project somehow. (And even then I would bet they just say: You are doing it wrong, fix your cflags like it happend with crosstool-ng )

Question to @borkra: What do you think? And are you still interested in that change?

borkra commented:
@chunkeey

Sure I will do that

borkra commented:
@chunkeey
> -	CFLAGS="-O2 $(filter-out -Os,$(call qstrip,$(TARGET_CFLAGS)))" \
> +	CFLAGS="-O2 $(filter-out -Os -mcpu=% -mfpu=%,$(call qstrip,$(TARGET_CFLAGS)))" \

This will be a problem,

because if we change architecture through -m flags glibc will not be compiled correctly with architectural settings. So -mcpu and -mfpu need to stay for glibc. There is no way glibc could know architecture we want it's not a psychic.

Although removing -m from gcc will resolve all problems at hand. So glibc changes is not really needed.

borkra commented:
@chunkeey 

I guess reading the manual helps.

I read this: https://gcc.gnu.org/install/configure.html

I will correct our gcc build process to comply, and it will resolve our issues.

fuqiang03 commented:
@chunkeey 
I think the problem with gcc8.3
Because other versions are normal.

borkra commented:
@fuqiang03

There are no problems, I am running all my routers  compiled with gcc 8 and do not have any problems.

borkra committed:
toolchain: Fixed option conflict during multilib compilation

This problem exposed when compiling glibc, but applicable across the board.
gcc compiles runtime libraries for all supported architectures, unless
otherwise specified, and later selects applicable library based -m options,
thus -m options should not be passed to gcc as they create broken multilib.

This change likely permit us to use single toolchain for all CPUs
in the family, reducing the number of toolchains needed

Signed-off-by: Boris Krasnovskiy <borkra@gmail.com>

borkra commented:
@chunkeey 

I committed the change  as I see it, please let me know if you have concerns about it.

   toolchain: Fixed option conflict during multilib compilation
    
    This problem exposed when compiling glibc, but applicable across the board
    gcc compiles runtime libraries for all supported architectures, unless
    otherwise specified, and later selects applicable library based -m options,
    thus -m options should not be passed to gcc as they create broken multilib.
    
    This change likely permit us to use single toolchain for all CPUs
    in the family, reducing the number of toolchains needed

fuqiang03 commented:
Source Code Added without Modification
![QQ截图20190321175013](https://user-images.githubusercontent.com/26851715/54747753-b74a6200-4c0a-11e9-8861-a19ef82ef7cc.png)
![QQ截图20190321175044](https://user-images.githubusercontent.com/26851715/54747756-b87b8f00-4c0a-11e9-98e8-e5137f57d2b2.png)
![QQ截图20190321184314](https://user-images.githubusercontent.com/26851715/54747758-b9acbc00-4c0a-11e9-8c36-c9c4091c37ea.png)

borkra commented:
@fuqiang03

 do not observe any of the issues you are reporting everything works just fine for me, even with your config.

Statement from you screenshot: "selected processor does not support 'xxxx' in Thumb mode"

Makes me believe that you running modified build. OpenWRT does not support or enable Thumb mode.

Fixing Thumb mode is out of scope of this.
I know it does not work and will notvworry about fixing it, as ut is not the mode OpenWRT support.

borkra commented:
@fuqiang03

In .config file you provided I see you are using musl:

> > CONFIG_USE_MUSL=y
> CONFIG_SSP_SUPPORT=y
> CONFIG_BINUTILS_VERSION_2_31_1=y
> CONFIG_BINUTILS_VERSION="2.31.1"
> # CONFIG_GCC_USE_EMBEDDED_PATH_REMAP is not set
> CONFIG_GCC_VERSION="7.4.0"
> **CONFIG_LIBC="musl"**
> CONFIG_TARGET_SUFFIX="muslgnueabi"

And yet, on screenshot above I see glibc. How come?

fuqiang03 commented:
[.config.zip](https://github.com/openwrt/openwrt/files/2992174/default.config.zip)
I'm sorry I sent the wrong document.
This is the configuration file for my re-experiment.

chunkeey commented:
@borkra 

I've edited the commits and added them to my staging tree
As of now, I don't know if this compiles or not so if there's an
obvious issue, I'll fix it silently and do a update again.

(I'll let this run on my CI and test it on the ARMs I have 
RPI (arm1176jzf-s), RPI3 (cortex-a53), IPQ806x (cortex-a15),
IPQ40xx (cortex-a7) and bcm53xx (cortex-a9, but no fpu!)

https://git.openwrt.org/?p=openwrt/staging/chunkeey.git;a=commit;h=e78a8e3dc111d9cadecf5a1828cfe7b03d16fbc8

https://git.openwrt.org/?p=openwrt/staging/chunkeey.git;a=commit;h=b01767b8d345da05e80d42b1d77be27705d69472

you'll notice that changed patch 2 to just to the filter-out on ARM.
The reason is that I'm too chicken to test this on all architectures
that openwrt supports (not to mention this needs to be tested on
gcc 5 and 7 versions as well).

borkra commented:
@chunkeey

You went much further then what I proposed. As long as we going this road, and using default compiler settings always. I would propose to add --disable-multilib flag to gcc compilation, it will reduce compilation time and possibly binary size.

fuqiang03 commented:
https://gcc.gnu.org/gcc-8/changes.html
Not quite understand

borkra commented:
@fuqiang03 I tried your latest .config and it compiles just fine for me no errors.
Did you change some component versions in your local stuff?

fuqiang03 commented:
No changes were made during the test.

fuqiang03 commented:
Can external factors cause these problems?

borkra commented:
Possibly, environment variables could do that, they would bleed into the build. and change stuff.

fuqiang03 commented:
I won't modify environment variables

borkra commented:
> I won't modify environment variables

I am not sure you understood what I am talking about. Environment variables are very common to modify.
Exported variables of any shell scripts becomes part of environment, in addition there startup shell scripts that  set environment variables.

chunkeey commented:
> You went much further then what I proposed. As long as we going this road, and using default compiler settings always. I would propose to add --disable-multilib flag to gcc compilation, it will reduce compilation time and possibly binary size.

I think this is already the case:

https://github.com/openwrt/openwrt/blob/master/toolchain/gcc/common.mk#L103

That line comes from a patch from 2010
https://github.com/openwrt/openwrt/commit/5149ed151d7ce6f4bdb14a7ad5611777afff1091

borkra commented:
Oh sorry, I have not noticed. In this case all these -m flags make no sense. As without multilib we cannot modify architecture at compile time, the needed libraries will not be available. And that is true for all targets not just ARM.

chunkeey commented:
@borkra 
This will be tons of fun. I tried to test this on all combinations however bit rot has set in on a few of the combinations. So, I guess this has to do. I've just pushed the patches and in classic fashion: "deal with some of them later... :wink:". (There's also the issue that XSCALE is set as the target CPU for all arm architectures in /tmp/.targetinfo, but it seems harmless because it only affects the main Makefile target). Anyway: I'm sorry that it took so long. Thank you very much for sticking around!

as for @fuqiang03 concerns: I left the default gcc for the arm toolchain at 7.4.0. Because yeah, there is some stuff left to do:

```
cc1plus: warning: -Wabi won't warn about anything [-Wabi]
cc1plus: note: -Wabi warns about differences from the most up-to-date ABI, which is also used by default
cc1plus: note: use e.g. -Wabi=11 to warn about changes from GCC 7
```

diizzyy commented:
@chunkeey @borkra 
So if I understand this correctly we now have to "santize" alll packages so they don't set mcpu and mfpu?

borkra commented:
@diizzyy 
-mcpu and -mfpu are set generically by target.mk And it is not really needed. We do not need to sanitize all packages, they generally architecture transparent.

Since we do not support multilib, thus cannot change architecture on the fly, because compiler and libc RTL compiled for only single architecture.

diizzyy commented:
Most are but some hardset such as libx264.
If we look at (lib)x264 https://code.videolan.org/videolan/x264/blob/master/configure#L934
This should be filtered/patched right unless your CPU is A8 (mvebu uses A9 for instance)?

borkra commented:
Yes that one chose to default not to compiler default but to something else, so yes special handling needed here. But that is the problem for this particular package.

borkra commented:
@diizzyy If you wonder how Yocto handled this issue you will find this patch:
http://cgit.openembedded.org/openembedded-core/tree/meta/recipes-multimedia/x264/x264/don-t-default-to-cortex-a9-with-neon.patch

diizzyy commented:
Thanks, I was about to do the same myself :-)

chunkeey commented:
Put please don't overdo it. libx264 is sort of "capped" at ARMv6 + NEON
https://code.videolan.org/videolan/x264/blob/master/common/cpu.c#L88 (no ARMv7 is defined).

borkra commented:
@chunkeey

Regarding XSCALE the problem is only for CPU family Makefile, which does not have Cpu type it subtargets do. I propose to put if statement around dump function and not print if SUBTARGET variable is empty and SUBTARGETS is not. 

Wabi is not really a problem it's just informing us that we are compiling foe ABI 12. But I will try to find where -Wabi coming from.

borkra commented:
@chunkeey 

The error below was fixed by the following gcc commit. We can patch it in if you would like:
https://github.com/gcc-mirror/gcc/commit/daa87973f7a00bf3bb81d0644dd60f4efb83bb65#diff-29290654ed15025b440a9f8c57d05aa0

> cc1plus: warning: -Wabi won't warn about anything [-Wabi]
> cc1plus: note: -Wabi warns about differences from the most up-to-date ABI, which is also used by default
> cc1plus: note: use e.g. -Wabi=11 to warn about changes from GCC 7

chunkeey commented:
> The error below was fixed by the following gcc commit. We can patch it in if you would like:
gcc-mirror/gcc@daa8797#diff-29290654ed15025b440a9f8c57d05aa0
>
>>    cc1plus: warning: -Wabi won't warn about anything [-Wabi]
>>    cc1plus: note: -Wabi warns about differences from the most up-to-date ABI, which is also used by default
>>    cc1plus: note: use e.g. -Wabi=11 to warn about changes from GCC 7

This seems to be the case. From what I can tell, it's only in devel, I guess if this was more of an issue, they would have backported it by now. I've applied said patch (there was some diff + reject)
but otherwise it does silence the warnings.

> Regarding XSCALE the problem is only for CPU family Makefile, which does not have Cpu type it subtargets do. I propose to put if statement around dump function and not print if SUBTARGET variable is empty and SUBTARGETS is not.

Yes, I came to the same conclusion: Having xscale in there does not hurt, but it looks rather odd on targets like brcm2708 (rpis). I guess your proposal would be fine. I though about just getting rid of the xscale thing in target.mk altogether and just print a CPU_TYPE when it's available:

---
diff --git a/include/target.mk b/include/target.mk
index 5346ea1a91..0ab76c3102 100644
--- a/include/target.mk
+++ b/include/target.mk
@@ -181,9 +181,6 @@ ifeq ($(DUMP),1)
     CPU_CFLAGS_pentium = -march=pentium-mmx
     CPU_CFLAGS_pentium4 = -march=pentium4
   endif
-  ifneq ($(findstring arm,$(ARCH)),)
-    CPU_TYPE ?= xscale
-  endif
   ifeq ($(ARCH),powerpc)
     CPU_CFLAGS_603e:=-mcpu=603e
     CPU_CFLAGS_8540:=-mcpu=8540
@@ -281,7 +278,7 @@ define BuildTargets/DumpCurrent
 	 echo 'Target-Features: $(FEATURES)'; \
 	 echo 'Target-Depends: $(DEPENDS)'; \
 	 echo 'Target-Optimization: $(if $(CFLAGS),$(CFLAGS),$(DEFAULT_CFLAGS))'; \
-	 echo 'CPU-Type: $(CPU_TYPE)$(if $(CPU_SUBTYPE),+$(CPU_SUBTYPE))'; \
+	$(if $(CPU_TYPE),echo 'CPU-Type: $(CPU_TYPE)$(if $(CPU_SUBTYPE),+$(CPU_SUBTYPE))';) \
 	 echo 'Linux-Version: $(LINUX_VERSION)'; \
 	 echo 'Linux-Release: $(LINUX_RELEASE)'; \
 	 echo 'Linux-Kernel-Arch: $(LINUX_KARCH)'; \
---
What I haven't checked is what happens now if a bad target is introduced with no CPU_TYPE.

But I guess this would need to be run by the mailing-list anyway.

borkra commented:
I agree default to XSCALE is odd, and not needed.
Although the problem are all family makefiles. Some of the printouts completely empty. There is no true architecture associated with them, why print target info for the file that does not have any?
